p2p/peer.go

```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
+// - Disconnected: peer was disconnected, unmarking as connected and evicting,
+//   and broadcasts a PeerStatusDown peer update.
+//
 // We track dialing and connected states independently. This allows us to accept
 // an inbound connection from a peer while the router is also dialing an
 // outbound connection to that same peer, which will cause the dialer to
 // eventually error (when attempting to mark the peer as connected). This also
 // avoids race conditions where multiple goroutines may end up dialing a peer if
 // an incoming connection was briefly accepted and disconnected while we were
 // also dialing.
-type peerManager struct {
+type PeerManager struct {
+	options PeerManagerOptions
+
 	mtx           sync.Mutex
 	store         *peerStore
 	dialing       map[NodeID]bool
 	connected     map[NodeID]bool
+	evicting      map[NodeID]bool
 	subscriptions map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
 }
 
-// newPeerManager creates a new peer manager.
-func newPeerManager(store *peerStore) *peerManager {
-	return &peerManager{
-		store:         store,
+// PeerManagerOptions specifies options for a PeerManager.
+type PeerManagerOptions struct {
+	// PersistentPeers are peers that we want to maintain persistent connections
+	// to. These will be scored higher than other peers, and if
+	// MaxConnectedUpgrade is non-zero any lower-scored peers will be evicted if
+	// necessary to make room for these.
+	PersistentPeers []NodeID
+
+	// MaxConnected is the maximum number of connected peers (inbound and
+	// outbound). 0 means no limit.
+	MaxConnected uint16
+
+	// MaxConnectedUpgrade is the maximum number of additional peer connections to
+	// use for upgrading to better-scored peers. 0 disables peer upgrading.
+	//
+	// For example, if we are already connected to MaxConnected peers, but we
+	// know or learn about better-scored peers (e.g. configured persistent
+	// peers) that we are not connected too, then we can probe these peers by
+	// using up to MaxConnectedUpgrade connections, and once connected evict the
+	// lowest-scored connected peers. This also works for inbound connections,
+	// i.e. if a higher-scored peer attempts to connect to us, we can accept
+	// the connection and evict a lower-scored peer.
+	MaxConnectedUpgrade uint16
+
+	// MinRetryTime is the minimum time to wait between retries. Retry times
+	// double for each retry, up to MaxRetryTime. 0 disables retries.
+	MinRetryTime time.Duration
+
+	// MaxRetryTime is the maximum time to wait between retries. 0 means
+	// no maximum, in which case the retry time will keep doubling.
+	MaxRetryTime time.Duration
+
+	// MaxRetryTimePersistent is the maximum time to wait between retries for
+	// peers listed in PersistentPeers. 0 uses MaxRetryTime instead.
+	MaxRetryTimePersistent time.Duration
+
+	// FuzzRetryTime is the upper bound of a random interval added to
+	// retry times, to avoid thundering herds. 0 disables fuzzing.
```

Comment:
Maybe use the "jitter" nomenclature here instead? Or maybe just reference https://en.wikipedia.org/wiki/Jitter or something like:

```
Clients will retry failed calls, by waiting a specific amount of time between consecutive retries. In order to avoid the thundering herd problem, jitter can be purposefully introduced, in order to break the synchronization across the clients thereby avoiding collisions.
```

Reply:
Good call -- it's easy to mix this up with fuzz testing.